src/index.ts

```diff
@@ -854,6 +863,88 @@ export abstract class PinejsClientCore<PinejsClient> {
 		}
 	}
 
+	private waitAsync(delayMs: number): Promise<void> {
+		return new Promise<void>((resolve) => {
+			setTimeout(resolve, delayMs);
+		});
+	}
+
+	protected async callWithDelay<T>(
+		fnCall: (attempts: number, isLastAttempt: boolean) => Promise<T>,
+		delayMs: number,
+		attempt: number,
+		final: boolean,
+		logAttempt: boolean,
+	): Promise<T> {
+		if (!delayMs) {
+			return await fnCall(attempt, final);
+		}
+
+		if (logAttempt) {
+			const msg = `Retrying balena api request in ${
+				delayMs / 1000
+			} seconds (attempt ${attempt})`;
+			console.log(msg);
+		}
+
+		await this.waitAsync(delayMs);
+
+		return fnCall(attempt, final);
+	}
+
+	protected async callWithRetry<T>(
+		fnCall: (attempt: number, isLastAttempt: boolean) => Promise<T>,
+		retryParameters?: RetryParameters,
+	): Promise<T> {
+		const canRetryDefaultHandler = (err: any) => {
+			const code = err?.statusCode ?? 0;
+			return code === 0 || code === 429 || (code >= 500 && code < 600);
+		};
```

Comment:
Can you move this out to module level please rather than having it potentially being recreated for each call